File: Courses/cps280s/Grader/Grader.pl
#!/usr/bin/perl

#Kyle Head
#22-September-2013
#Grader.pl
# Grader.pl addes up the scores for every person in a class and then converts
# Their scores to a letter grade. The program accepts the first argument given
# as a file to be parsed. Given that the file is in the format specified in the
# homework it will add up the numbers given into a total grade. This grade will
# then be converted to a letter grade. The program then outputs to the Graded
# file in the format specified.  

my @names;
my $possible = 440;

open my $fh, "<", $ARGV[0]  || die "Can't open $argv[0]: $!";
open my $graded, ">", "Graded" || die "Can't open Graded: $!";
my @grades = <$fh>;

my @grades = sort {lc($a) cmp lc($b)} @grades;
@totals = map {
  my @f = split ",";
  $name = join " ", $f[1], $f[0];
  push @names, $name;
  my $total;
  $total += $_ for splice @f, 2;
  $total = $total / $possible * 100;
  $total;
} @grades;
@totals = map {
	my $grade;
	if ($_ >= 92 && $_ <= 100)
	{
		$grade = "A";
	}
	elsif ($_ >= 90 && $_ < 92)
	{
		$grade = "A-";
	}
	elsif ($_ >= 88 && $_ < 90)
	{
		$grade = "B+";
	}
	elsif ($_ >= 82 && $_ < 88)
	{
		$grade = "B";
	}
	elsif ($_ >= 80 && $_ < 82)
	{
		$grade = "B-";
	}
	elsif ($_ >= 78 && $_ < 80)
	{
		$grade = "C+";
	}
	elsif ($_ >= 72 && $_ < 78)
	{
		$grade = "C";
	}
	elsif ($_ >= 70 && $_ < 72)
	{
		$grade = "C-";
	}
	elsif ($_ >= 68 && $_ < 70)
	{
		$grade = "D+";
	}
	elsif ($_ >= 62 && $_ < 68)
	{
		$grade = "D";
	}
	elsif ($_ >= 60 && $_ < 62)
	{
		$grade = "D-";
	}
	else
	{
		$grade = "E";
	}
	$grade;
} @totals;
for(my $i = 0; $i <= ($#totals); $i++){
	print { $graded } "$names[$i]: $totals[$i]\n";
}
close $graded || die "Can't close Graded: $!";

File: Courses/cps280s/PyGrader/Grader.py
'''
Created on Nov 12, 2013

@author: Kyle Head
Implements the Grader program in Python. Program takes in grades and names
from the Grades file and then totals the grades into a letter grade. The 
new letter grade is then written to the Graded file along with the names of
the students.
'''

def main():
    possible = 400
    points = []
    graded = []
    try:
        lines = [line.strip() for line in open('Grades','r')]
    except IOError as e:
        print "Cannot write to Grades file!".format(e.errno, e.strerror)
    lines = sorted(lines, key=lambda name: name[0].lower())
    for i in range(len(lines)):
        temp= lines[i].split(',')
        graded.append([temp[1], temp[0]])
        total = 0
        for scores in temp[2:-1]:
            total += int(scores)
        points.append(100 * (float(total)/float(possible)))
    for i in range(len(points)):
        graded[i].append(points[i])

    for i in range(len(graded)): 
        if (graded[i][2] >= 92):
            graded[i][2] = "A"
        elif (graded[i][2] >= 90 and graded[i][2] < 92):
            graded[i][2] = "A-"
        elif (graded[i][2] >= 88 and graded[i][2] < 90):
            graded[i][2] = "B+"
        elif (graded[i][2] >= 82 and graded[i][2] < 88):
            graded[i][2] = "B"
        elif (graded[i][2] >= 80 and graded[i][2] < 82):
            graded[i][2] = "B-"
        elif (graded[i][2] >= 78 and graded[i][2] < 80):
            graded[i][2] = "C+"
        elif (graded[i][2] >= 72 and graded[i][2] < 78):
            graded[i][2] = "C"
        elif (graded[i][2] >= 70 and graded[i][2] < 72):
            graded[i][2] = "C-"
        elif (graded[i][2] >= 68 and graded[i][2] < 70):
            graded[i][2] = "D+"
        elif (graded[i][2] >= 62 and graded[i][2] < 68):
            graded[i][2] = "D"
        elif (graded[i][2] >= 60 and graded[i][2] < 62):
            graded[i][2] = "D-"
        else:
            graded[i][2] = "E"
    f =open('Graded', 'w')
    for item in graded:
        f.write("{0} {1}:{2}\n".format(item[0], item[1], item[2]))
    
if __name__ == '__main__':
    main()